### ISSUE TITLE
Import failing with "Cannot read property '0' of undefined"

### ISSUE THREAD

**Issue by lddubeau** [systemjs/systemjs#1640]
### Versions

I get the problem with versions 0.20.0 to 0.20.11.

I was not able to reproduce it with any version in the 0.19 series.

### Symptoms

I get a big fat exception:

```
Error: Cannot read property '0' of undefined
```

Which points to [this](https://github.com/systemjs/systemjs/blob/0.20.10/dist/system.src.js#L1029). It is the `var depLoad = dependencyInstantiations[i];` line in the following code:

```
function makeDynamicRequire (loader, key, dependencies, dependencyInstantiations, registry, state, seen) {
  // we can only require from already-known dependencies
  return function (name) {
    for (var i = 0; i < dependencies.length; i++) {
      if (dependencies[i] === name) {
        var depLoad = dependencyInstantiations[i];
        var module;

        if (depLoad instanceof ModuleNamespace)
          module = depLoad;
```

When I add tracing to the console, I find that `makeDynamicRequire` is sometimes called with a `dependencies` array which has a length > 0 but with `dependencyInstantiations` that is undefined.

### How to reproduce the problem

Clone https://github.com/lddubeau/systemjs_issue3

1. ``npm install``.

2. Start the server ``./server.js localhost:[some port number]``. **It
is important that you use THIS SERVER. The problem here is a race
condition, and the server purposely delays responding to the request
for one module so as to recreate the conditions necessary for the
problem to occur.**

3. Direct your browser to load ``http://localhost:[same port number]/index.html``.

If you are lucky, you'll immediately get:

```
Uncaught (in promise) Error: Cannot read property '0' of undefined
  Loading x
    at register-loader.js:532
    at doEvaluate (register-loader.js:607)
    at ensureEvaluate (register-loader.js:518)
    at register-loader.js:538
    at doEvaluate (register-loader.js:607)
    at ensureEvaluate (register-loader.js:518)
    at register-loader.js:538
    at doEvaluate (register-loader.js:607)
    at ensureEvaluate (register-loader.js:518)
    at register-loader.js:538
```

If you are unlucky, you may need to reload until you get the error, or
it might require that you play with the timeout that is set in
index.html.

When I try it here I get the error nearly 100% of the time.

**Comment by lddubeau** [comment]
I'm hypothesizing this problem is caused by the way `instantiateDeps` sometimes "truncates" promises when it deems a dependency has having been "seen" already. Let's consider a simple case:

A depends on B, which depends on C, which depends on D.

A is imported. By the time `instantiateDeps` is done with the whole shebang, the `link.depsInstantiatePromise` for A will be a promise that resolves when all instantiations for A, B, C and D are done. The `link.depsInstantiatePromise` for B resolves when all instantiations for B, C and D are done, etc.

Now, suppose the following dependencies:

A depends on C and B.
B depends on B2.
B2 depends on C.
C depends on D.

Let's suppose `SystemJS.import("B")`, and no other calls. This is the same as the case above: `link.depsInstantiatePromise` for each module will cover the module and the dependencies of the module.

Now let's suppose `SystemJS.import("A")`. By the time `instantiateDeps` gets to computing the `deepDepsInstantiatePromises` array for B2, C will have already been seen. So the `link.depsInstantiatePromise` for B2 is "truncated" comparatively to the scenario above.

If we only have `SystemJS.import("A")` and nothing else, the truncation won't matter.  The problem occurs if another concurrent import needs the same modules. If a concurrent import needs B2 (directly or through B), it may rely on the "truncated" `link.depsInstantiatePromise` what was set by the import of A.

I'll note that if I change the [logic for handling seen modules](https://github.com/ModuleLoader/es-module-loader/blob/7abed12f4549c274c6fcbce68b7e80002a10f73b/core/register-loader.js#L409) to the following, then I no longer get an error:

```
      if (seen.indexOf(depLoad) !== -1) {
        if (!depLink.depsInstantiatePromise) {
          throw new Error("missing promise in custom code");
        }
        console.warn("promise substitution for", depLoad);
        deepDepsInstantiatePromises.push(depLink.depsInstantiatePromise);
        continue;
      }
      seen.push(depLoad);
```

The logic I went by is that if it has been seen, then it necessarily must have a promise that is already set, and we can wait for that instead of waiting for nothing.

(Oops... I just realized while adding the link above that the offending code is actually in [es-module-loader](https://github.com/ModuleLoader/es-module-loader).)

**Comment by tflori** [comment]
I get the same error - maybe the case is differnt. I found out it was a problem with a class that extends a class in a directory above:

```
- src
  TranslationLoader.ts # export abstract class TranslationLoader
  - TranslationLoader
    Json.ts # import {TranslationLoader} from "../TranslationLoader"
```

I don't understand this problem really but magically sometimes it works but never when I run debugger. And because I need debugging right know I'm trying since 3 hours why it is not working.

It is annoying that I cant put this class to subdirectory but for now it seems a working solution for me...

**Comment by tflori** [comment]
sorry It's wrong. I can't figure out what is the problem. sometimes it works but mostly I get this error not:

```
TypeError: Cannot read property '0' of undefined
    at http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?503c55c7d553c659718f46300e06b9fbd1020ca4:1034:53
```

I don't change anything also I don't recompile - I just start karma by `node_modules/.bin/karma start --browsers Chrome`. It does not change when I use Firefox.

**Comment by tflori** [comment]
I found since when I have the problem: since I added a dependency to a class. The point is: it is no circular dependency.

TranslatorContainer requires TranslatorConfig and TranslatorConfig requires TranslationLoader and TranslationLoaderJson. TranslationLoaderJson requires TranslationLoader but both don't require TranslatorContainer or TranslatorConfig. Also this problem came with TranslatorContainer requires TranslatorConfig. TranslatorConfig existed before and it wasn't any problem.

At the end I found this problem: sometimes `makeDynamicRequire` gets called before `Promise.all(depsInstantiatePromises)` from `instantiateDeps` gets resolved. Also I don't understand why it happends much more rarely when I add breakpoints there. It feels like a problem of Heisenbergsche Unschärferelation (Uncertainty principle).
